src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.Unix.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>This class contains methods that are mainly used to manage native memory.</summary>
+    public static unsafe class NativeMemory
+    {
+        /// <summary>Allocates an aligned block of memory of the specified size and alignment, in bytes.</summary>
+        /// <param name="alignment">The alignment, in bytes, of the block to allocate.</param>
+        /// <param name="byteCount">The size, in bytes, of the block to allocate.</param>
+        /// <returns>A pointer to the allocated aligned block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="byteCount" /> of memory failed.</exception>
+        /// <remarks>
+        ///     <para>This method converts a <paramref name="byteCount" /> of <c>0</c> to <c>1</c>.</para>
+        ///     <para>This method is a thin wrapper over the C <c>aligned_alloc</c> API or a platform dependent aligned allocation API such as <c>_aligned_malloc</c> on Win32.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AlignedAlloc(nuint alignment, nuint byteCount)
```

Comment:
The C API doesn't strictly restrict this to `alignment == power_of_two`, but many impls do.

We could check or we could just pass-through and recommend users to pass in powers of two.

Reply:
We should have predictable and diagnosable behavior for invalid argument.

What is going to be the behavior when somebody passes in alignment that is not power of two? Are people going to be able to figure out what they have done wrong?

Reply:
For `aligned_alloc`
> Passing a size which is not an integral multiple of alignment or a alignment which is not valid or not supported by the implementation causes the function to fail and return a null pointer (C11, as published, specified undefined behavior in this case, this was corrected by DR 460).

> This function sets errno to ENOMEM if the memory allocation failed or if the requested size was greater than _HEAP_MAXREQ. For more information about errno, see errno, _doserrno, _sys_errlist, and _sys_nerr. Also, _aligned_malloc validates its parameters. If alignment is not a power of 2 or size is zero, this function invokes the invalid parameter handler, as described in Parameter Validation. If execution is allowed to continue, this function returns NULL and sets errno to EINVAL.

So in both cases, it should return `null` and treat it as an OOM. The exact alignments supported must be powers of two on windows and can potentially be something else if the C function is called instead

Reply:
>  this function invokes the invalid parameter handler, as described in Parameter Validation. 

If this going to lead to fail-fast when somebody configures the invalid pararameter handler and passes invalid alignment? That is not the behavior one would expect from .NET API.

Reply:
Allowing a invalid parameter handler is the case for many CRT functions on Windows, and our concerns are far beyond this API if a user hooks up one and decides to "fail fast": https://docs.microsoft.com/en-us/cpp/c-runtime-library/parameter-validation?view=msvc-160

Reply:
We always validate the arguments upfront to avoid hitting the CRT invalid parameter handler. We treat invalid parameter handler the same way as asserts. If you can hit it by calling .NET API, we consider it to be a bug in .NET.

> By default, the invalid parameter calls _invoke_watson,

I have not realized that it is fail fast by default...

Reply:
I'll add the explicit validation for `_aligned_alloc` then to avoid this scenario.